README.md
# OpenSshCC ("OpenSSH CC Toggle")
A Control Center toggle for turning the SSH server on/off on (jailbroken) iOS.

This used to be an all-in-one tweak called `SSHonCC` ("SSH Toggle and Port") and it had the [OpenSSH Settings](https://gitlab.com/blanxd/OpenSshPort) and [SSHswitch](https://gitlab.com/blanxd/SSHswitch) as its subprojects. It was broken up into separate packages for everyone to have a choice about some dependencies et al, most of the development really happens in `SSHswitch` anyway. It started in 2018, since there wasn't anything similar for iOS11 long after it got publicly jailbroken, like there used to be on earlier jailbreaks. And the 1st public jb for iOS11 had OpenSSH server forcibly installed without an option to turn it off so to me it actually felt like a security issue, because not everyone knows what it is in the 1st place so most were probably walking around listening for anyone to log in on the default port with the default password.

## Building
- By default this Makefile builds it for rootless jbs (since OpenSshCC ver.1.1.0), with the default toolchain on the system. To build for rootful environments and/or use the old ABI for arm64e, append RF=1 to the make commandline, eg. `make RF=1` or `make package FINALPACKAGE=1 RF=1`, in this case it tries to find an Xcode 11 toolchain for the job (don't forget to `make clean` between RF and RL).  
- A recent [Theos](https://theos.dev/) is recommended (post 2023-03-26), but it also does the rootless job with older Theos versions.  

## Changelog
The [Changelog.md](Changelog.md) shows how it has evolved, it is mostly about what changed before the projects were broken up, the CC Toggle itself has really always had one simple function.  


Changelog.md
# Changelog of OpenSshCC (u.blanxd.opensshcc)
**CHANGES 1.1.0 [2023-07-20]**
* Added rootless support
* Added compatibility with SSHswitch 1.1.0

**CHANGES 1.0.1 [2021-03-04]**
* No functional changes, some packaging improvements.

**CHANGES 1.0 [2020-12-31] (compared to SSHonCC v.1.4 where this was the master project)**
* No functional changes compared to the last few SSHonCC releases. Break-up release.

### SSHonCC changelog
**SSHonCC CHANGES 1.4 [2020-05-27]**
* Added option to turn off Password Authentication.
* Lots of internal optimizations (no more C++).
* Added some functionality to SSHswitch which might be used in the future.

**SSHonCC CHANGES 1.3.5 [2019-11-25]**
* Same as 1.3.3/1.3.4, with Checkra1n and iOS 13 compatibility.

**SSHonCC CHANGES 1.3.4 [2019-04-30]**
* Same as 1.3.3, added support for A12 devices.

**SSHonCC CHANGES 1.3.3 [2018-10-20]**
* Added the ability to change the port(s) if ssh server is not running as sshd daemon, but via launchd listener (ie. compatibility with unc0ver).

**SSHonCC CHANGES 1.3.2 [2018-09-08]**
* Incoming scp and sftp connections now also get disconnected when turning sshd off (unless specifically allowed to persist).
* Internal optimizations, not dependent on coreutils any more.

**SSHonCC CHANGES 1.3.1 [2018-08-26]**
* Enabled specifying several ports for sshd to listen on.
* Internal stability fixes (none reported, but in theory v.1.3 was capable of crashing a few things).

**SSHonCC CHANGES 1.3 [2018-08-21] (removed, see 1.3.1 changelog)**
* Added options to control whether sshd starts (or doesn't) after rebooting/re-jailbreaking, despite the running toggled state.

**SSHonCC CHANGES 1.2 [2018-07-23]**
* Added option to disallow root logins.

**SSHonCC CHANGES 1.1 [2018-05-28]**
* When turning sshd off, active sessions now get disconnected. This functionality can be turned off in Settings.

**SSHonCC CHANGES 1.0.1 [2018-05-17]**
* packaging improvements, so in case it's uninstalled, sshd gets turned on with default port.

**SSHonCC CHANGES 1.0 [2018-05-13]**
* Initial release


SSHonCC.m
#import "SSHonCC.h"

#import <objc/runtime.h> // objc_getClass()
#include <stdio.h> // snprintf, vsprintf, fopen, fgets etc..
#include <sys/types.h> // pid_t
#include <spawn.h> // posix_spawn and friends
#include <sys/wait.h> // waitpid
#include <string.h> // strerror
#include <sys/stat.h>  // stat
#include <stdarg.h> // for my log func

// This probably isn't sending anything on arm64e devices on iOS14 if built with Xcode11
static void 
bhLog(NSString *format, ...){
	va_list argp;
	va_start(argp, format);
	NSLogv([NSString stringWithFormat:@"%.4f u.blanxd.opensshcc %@", ([[NSDate date] timeIntervalSince1970] * 1000.0)/1000, format], argp);
	va_end(argp);
}

@interface UIImage ()
+ (UIImage *)imageNamed:(NSString *)name inBundle:(NSBundle *)bundle;
@end

@interface FBSystemApp : UIApplication
@end

@interface SBUserAgent : NSObject
-(BOOL)deviceIsPasscodeLocked;
-(BOOL)deviceIsLocked;
-(BOOL)openURL:(id)arg1 allowUnlock:(BOOL)arg2 animated:(BOOL)arg3;
@end

@interface SpringBoard: FBSystemApp <UIApplicationDelegate>
-(SBUserAgent *)pluginUserAgent;
@end

/* so this seems to only work starting iOS 13. For 11,12 still need to attach the GestureRecognizer straight to the ToggleViewController  */
@interface CCUIToggleViewController (SSHonCC)
-(BOOL)shouldFinishTransitionToExpandedContentModule;
@end

@implementation CCUIToggleViewController (SSHonCC)
-(BOOL)shouldFinishTransitionToExpandedContentModule {

	if ( [[self module] isKindOfClass:[SSHonCC class]] )
		[(SSHonCC*)[self module] openSshSettings];

	return NO;
}
@end


@implementation SSHonCC

@synthesize iconGlyph; 
@synthesize selectedIconGlyph;
@synthesize selectedColor;

-(SSHonCC *) init {
    self = [super init];

	struct stat ftest;
	
	// mitigating XinaA15's overwriting procedures
	char * usr = "usr";
	char * bin = "bin";
	char * ssw = "SSHswitch";
	_p_sshswitch = [NSString stringWithFormat:@"/%s/%s/%s", usr, bin, ssw]; // rooted
	if ( stat([_p_sshswitch UTF8String], &ftest) != 0 ){
		_p_sshswitch = [NSString stringWithFormat:@"/var/jb%@", _p_sshswitch]; // simple /var/jb
		if ( stat([_p_sshswitch UTF8String], &ftest) != 0 ){
			NSString * envJbRoot = [[[NSProcessInfo processInfo] environment] objectForKey:@"JB_ROOT_PATH"];
			if ( envJbRoot ){
				_p_sshswitch = [NSString stringWithFormat:@"%@/%s/%s/%s", envJbRoot, usr, bin, ssw];
			} // else now we're f*d, this isn't going to work
		}
	}

	_openSshIsRunning = -1;

	// in 11,12 we can access the super.contentViewController.view
	// in 13 we cannot I guess, it crashes SB
	/* doing it in the Category only works starting iOS 13 */
	if ( ![self respondsToSelector:@selector(contentViewControllerForContext:)] ){
		// 11,12
		[super.contentViewController.view addGestureRecognizer:[[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(openSshSettings)]];
	}

	_longPressOnce = NO;
	_didThatMyself = NO;
	
	CFNotificationCenterAddObserver(
		CFNotificationCenterGetDarwinNotifyCenter(), 
		(__bridge const void *)(self),
		openSshOnOffCallback, 
		CFSTR("com.openssh.sshd/on"), 
		NULL, 
		CFNotificationSuspensionBehaviorCoalesce
		);
	CFNotificationCenterAddObserver(
		CFNotificationCenterGetDarwinNotifyCenter(), 
		(__bridge const void *)(self),
		openSshOnOffCallback, 
		CFSTR("com.openssh.sshd/off"), 
		NULL, 
		CFNotificationSuspensionBehaviorCoalesce
		);

	
	return self;
}

// ideas from https://stackoverflow.com/questions/26637023/how-to-properly-use-cfnotificationcenteraddobserver-in-swift-for-ios
- (void)notificationCallbackReceivedWithName:(NSString *)name {
	if ( _didThatMyself ){
		_didThatMyself = NO;
		return;
	}	
	_openSshIsRunning = [name isEqualToString:@"com.openssh.sshd/on"] ? 1 : 0;
	[self refreshState];
}

// so this isn't really a part of the class.
static void 
openSshOnOffCallback(CFNotificationCenterRef center, void *observer, CFStringRef name, const void *object, CFDictionaryRef userInfo) {
	[(__bridge SSHonCC *)observer notificationCallbackReceivedWithName:(__bridge NSString *)name];		
}

- (void)openSshSettings {
	if ( !_longPressOnce ){
		Class SpringBoardClass = objc_getClass("SpringBoard");
		if ( ![[(SpringBoard *)[SpringBoardClass sharedApplication] pluginUserAgent] deviceIsLocked] ){
			_longPressOnce = YES;
			NSString *prefUrl = @"prefs:root=OpenSSH";
			if([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:prefUrl]]){
				BOOL success = [[(SpringBoard *)[SpringBoardClass sharedApplication] pluginUserAgent] openURL:[NSURL URLWithString:prefUrl] allowUnlock:YES animated:NO];
	    	   	if (!success) {
   			        bhLog(@"longPress FAILED Open %@", prefUrl);
	    	    }
			} else {
				bhLog(@"longPress cannot Open %@", prefUrl);
			}
			[self performSelector:@selector(longPressDone) withObject:nil afterDelay:2];
		}
	}
}
- (void)longPressDone {
	_longPressOnce = NO;
}

- (UIImage *)iconGlyph {
	if ( iconGlyph==nil )
		iconGlyph = [UIImage imageNamed:@"Icon" inBundle:[NSBundle bundleForClass:[self class]]];
	return iconGlyph;
}

-(UIImage *)selectedIconGlyph {
	if ( selectedIconGlyph == nil )
		selectedIconGlyph = [self iconGlyph];
	return selectedIconGlyph;
}


- (UIColor *)selectedColor {
	if ( selectedColor == nil )
		selectedColor = [UIColor blackColor];
	return selectedColor;
}

- (BOOL)isSelected {
	BOOL sel;
	if ( _openSshIsRunning > -1 )
		sel = _openSshIsRunning > 0;
	else
		sel = [self runSwitch:-1];
	_openSshIsRunning = -1;
	return sel;
}

- (void)setSelected:(BOOL)selected {

	if ( _longPressOnce )
		return;

	_didThatMyself = YES;

	Class SpringBoardClass = objc_getClass("SpringBoard");

	if ( ![[(SpringBoard *)[SpringBoardClass sharedApplication] pluginUserAgent] deviceIsPasscodeLocked] || [self evenIfLocked] ){

		if ( ![self runSwitch: selected ? 1 : 0] )
			return;
		[super refreshState];

	}
	
} // - (void)setSelected:(BOOL)selected {

- (BOOL)evenIfLocked {

	char * fp_tweakpref = "/var/mobile/Library/Preferences/u.blanxd.sshswitch/prefs";

	FILE * ff_tweakpref;
	ff_tweakpref = fopen(fp_tweakpref, "r");
	if ( ff_tweakpref != NULL ){
		_Bool toolong = 0;
		char istr[64];
		while ( fgets(istr,64,ff_tweakpref) != NULL ){
			if ( strpbrk(istr,"\n")==NULL ){
				toolong = 1; // skip the next chunks until the line ends
				continue;
			} else if ( toolong==1 ){
				toolong = 0; // next one could be ok
				continue;
			}
	
			if ( strlen(istr) < 2 )
				continue;
			
			if ( istr[0] == 'e' ){
				fclose(ff_tweakpref);
				return YES;
			}
		}
		fclose(ff_tweakpref);
	}

	struct stat elockedfs;
	if ( stat( "/etc/ssh/u.blanxd.sshswitch.eveniflocked", &elockedfs ) == 0 )
		return YES;
	
	return NO;
}

// onOff: -1=get status: 1=running, 0=not
// onOff: 0=turn off
// onOff: 1=turn on
- (BOOL)runSwitch:(int8_t)onOff {

	int status = 2;
	if ( onOff>=0 )
		_didThatMyself = YES;

	posix_spawn_file_actions_t action;
	posix_spawn_file_actions_init(&action);
	posix_spawn_file_actions_addopen (&action, STDIN_FILENO, "/dev/null", O_RDONLY, 0);
	posix_spawn_file_actions_addopen (&action, STDOUT_FILENO, "/dev/null", O_RDONLY, 0);
	posix_spawn_file_actions_addopen (&action, STDERR_FILENO, "/dev/null", O_RDONLY, 0);

	uint8_t ai = onOff<0 ? 0 : 1;
	const char *args[ 2+ai ];

	args[0] = "SSHswitch";
	if ( ai>0 ) args[1] = onOff ? "on" : "off";
	args[++ai] = NULL;

	pid_t pid;
	status = posix_spawn(&pid, [_p_sshswitch UTF8String], &action, NULL, (char* const*)args, NULL);
	if (status == 0) {
		if (waitpid(pid, &status, 0) != -1) {
			if ( WIFEXITED(status) != 0 ){
				status = WEXITSTATUS(status);
			} else {
				bhLog(@"runSwitch(%d) ERROR: WIFEXITED(status) = 0 (SSHswitch pid:%d)", onOff, pid);
				status = 2;
			}
		} else {
			bhLog(@"runSwitch(%d) ERROR: waitpid(%d,&status,0) = -1", onOff, pid);
			status = 2;
		}
	} else {
		bhLog(@"runSwitch(%d) ERROR: posix_spawn() = %d - %s", onOff, status, strerror(status));
		status = 2;
	}
	posix_spawn_file_actions_destroy(&action);

	if ( onOff<0 ){
		_openSshIsRunning = status==0 ? 1 : 0;
	} else if ( status==0 ){
		_openSshIsRunning = onOff>0 ? 1 : 0;
		return YES;
	} else
		return NO;

	return _openSshIsRunning>0;

} // - (BOOL)runSwitch:(int8_t)onOff {

@end


headers/ControlCenterUIKit/CCUIToggleViewController.h
/* This file is from "the internet", well I've combined 3 different versions over time.
 * Many things have been commented out since they're not needed for this project
 * and might need a bunch of other header files.
 * Things not included with the project should be available be default in Theos (namely ControlCenterUIKit).
 * Blanxd.H
 */

/*
* This header is generated by classdump-dyld 0.1
* on Wednesday, September 20, 2017 at 10:02:12 PM Eastern European Summer Time
* Operating System: Version 11.0 (Build 15A372)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is free of use, Copyright © 2013 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Friday, February 8, 2019 at 1:25:48 PM Eastern European Standard Time
* Operating System: Version 12.1 (Build 16B92)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Sunday, November 10, 2019 at 11:16:35 PM Eastern European Standard Time
* Operating System: Version 13.1.3 (Build 17A878)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

#import <ControlCenterUIKit/CCUIButtonModuleViewController.h>
//#import <ControlCenterUIKit/CCUIContentModuleContentViewController.h>
#import <ControlCenterUIKit/CCUIContentModuleContentViewController-Protocol.h>
//#import <ControlCenterUIKit/CCUIToggleModule.h>

@class UIImageView, UIImage, UIColor, CCUICAPackageDescription, NSString, CCUIToggleModule;
//, UIViewPropertyAnimator; // 13

@interface CCUIToggleViewController : CCUIButtonModuleViewController <CCUIContentModuleContentViewController> {
	UIImageView* _glyphImageView;
	UIImage* _glyphImage;
	UIImage* _selectedGlyphImage;
	UIColor* _selectedColor;
	CCUICAPackageDescription* _glyphPackageDescription;
	NSString* _glyphState;
	CCUIToggleModule* _module;
}
@property (assign,nonatomic) CCUIToggleModule* module;                   //@synthesize module=_module - In the implementation block
//@property (assign,nonatomic,__weak) CCUIToggleModule * module;                              //@synthesize module=_module - In the implementation block
//@property (readonly) unsigned long hash; // 11
//@property (readonly) unsigned long long hash; // 12,13
//@property (readonly) Class superclass; 
//@property (copy,readonly) NSString * description; 
//@property (copy,readonly) NSString * debugDescription; 
@property (nonatomic,readonly) double preferredExpandedContentHeight; 
@property (nonatomic,readonly) double preferredExpandedContentWidth; 
@property (nonatomic,readonly) double preferredExpandedContinuousCornerRadius; // 13
@property (nonatomic,readonly) BOOL providesOwnPlatter; 
//@property (nonatomic,readonly) UIViewPropertyAnimator * customAnimator; // 13
-(void)viewWillAppear:(BOOL)arg1 ;
-(void)viewDidLoad;
-(void)refreshState;
-(BOOL)_canShowWhileLocked; // 13
//-(id)module; // 11
-(CCUIToggleModule *)module;
//-(void)setModule:(id)arg1 ; // 11
-(void)setModule:(CCUIToggleModule *)arg1 ;
-(void)buttonTapped:(id)arg1 forEvent:(id)arg2 ;
-(double)preferredExpandedContentHeight;
-(BOOL)shouldFinishTransitionToExpandedContentModule;
-(void)reconfigureView;
@end


headers/ControlCenterUIKit/CCUIButtonModuleViewController.h
/* This file is from "the internet", well I've combined 3 different versions over time.
 * Many things have been commented out since they're not needed for this project
 * and might need a bunch of other header files.
 * Things not included with the project should be available be default in Theos (namely ControlCenterUIKit).
 * Blanxd.H
 */

/*
* This header is generated by classdump-dyld 1.0
* on Thursday, January 25, 2018 at 11:27:41 PM Eastern European Standard Time
* Operating System: Version 11.1.2 (Build 15B202)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Friday, February 8, 2019 at 1:25:48 PM Eastern European Standard Time
* Operating System: Version 12.1 (Build 16B92)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Sunday, November 10, 2019 at 11:16:35 PM Eastern European Standard Time
* Operating System: Version 13.1.3 (Build 17A878)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

#import <UIKit/UIViewController.h> // 11 // ?? 12,13 ??
//#import <UIKitCore/UIViewController.h> // 12,13
//#import <libobjc.A.dylib/CCUIContentModuleContentViewController.h>
#import <ControlCenterUIKit/CCUIContentModuleContentViewController-Protocol.h>

//@class UIViewPropertyAnimator, // 13
@class CCUIButtonModuleView, UIImage, UIColor; //, CCUICAPackageDescription, NSString;

@interface CCUIButtonModuleViewController : UIViewController <CCUIContentModuleContentViewController> //{
//	CCUIButtonModuleView* _buttonModuleView;
//	BOOL _expanded;
//}
@property (nonatomic,retain) UIImage * glyphImage; 
@property (nonatomic,retain) UIColor * glyphColor; 
@property (nonatomic,retain) UIImage * selectedGlyphImage; 
@property (nonatomic,retain) UIColor * selectedGlyphColor; 
//@property (nonatomic,retain) CCUICAPackageDescription * glyphPackageDescription; 
@property (nonatomic,copy) NSString * glyphState; 
@property (assign,nonatomic) double glyphScale; // 13
@property (assign,getter=isSelected,nonatomic) BOOL selected; 
//@property (getter=isExpanded,nonatomic,readonly) BOOL expanded; // 11,12                              //@synthesize expanded=_expanded - In the implementation block
@property (assign,getter=isExpanded,nonatomic) BOOL expanded;                                 //@synthesize expanded=_expanded - In the implementation block
@property (nonatomic,readonly) CCUIButtonModuleView * buttonView; 
@property (nonatomic,readonly) BOOL hasGlyph; // 13
//@property (readonly) unsigned long long hash; 
//@property (readonly) Class superclass; 
//@property (copy,readonly) NSString * description; 
//@property (copy,readonly) NSString * debugDescription; 
@property (nonatomic,readonly) double preferredExpandedContentHeight; 
@property (nonatomic,readonly) double preferredExpandedContentWidth; 
@property (nonatomic,readonly) double preferredExpandedContinuousCornerRadius; // 13
@property (nonatomic,readonly) BOOL providesOwnPlatter; 
@property (nonatomic,readonly) UIViewPropertyAnimator * customAnimator; // 13
-(BOOL)isSelected;
-(void)viewDidLoad;
-(void)viewWillLayoutSubviews; // 13
-(void)setSelected:(BOOL)arg1 ;
-(BOOL)isExpanded;
-(void)setExpanded:(BOOL)arg1 ; // 13
-(UIColor *)selectedGlyphColor;
-(void)setSelectedGlyphColor:(UIColor *)arg1 ;
-(void)setGlyphColor:(UIColor *)arg1 ;
-(UIColor *)glyphColor;
-(UIImage *)glyphImage;
-(void)setGlyphImage:(UIImage *)arg1 ;
-(void)setSelectedGlyphImage:(UIImage *)arg1 ;
-(UIImage *)selectedGlyphImage;
-(CCUIButtonModuleView *)buttonView;
//-(void)setGlyphPackageDescription:(CCUICAPackageDescription *)arg1 ;
-(void)setGlyphState:(NSString *)arg1 ;
//-(CCUICAPackageDescription *)glyphPackageDescription;
-(NSString *)glyphState;
-(void)setGlyphScale:(double)arg1 ; // 13
-(double)glyphScale; // 13
-(void)buttonTouchDown:(id)arg1 forEvent:(id)arg2 ; // 13
-(void)buttonTapped:(id)arg1 forEvent:(id)arg2 ;
-(void)_buttonTouchDown:(id)arg1 forEvent:(id)arg2 ; // 13
-(void)_buttonTapped:(id)arg1 forEvent:(id)arg2 ; // 12,13
-(double)preferredExpandedContentHeight;
-(void)willTransitionToExpandedContentMode:(BOOL)arg1 ;
-(void)didTransitionToExpandedContentMode:(BOOL)arg1 ;
-(BOOL)hasGlyph; // 13
@end


headers/ControlCenterUIKit/CCUIButtonModuleView.h
/* This file is from "the internet", well I've combined 3 different versions over time.
 * Many things have been commented out since they're not needed for this project
 * and might need a bunch of other header files.
 * Blanxd.H
 */
 
/*
* This header is generated by classdump-dyld 1.0
* on Thursday, January 25, 2018 at 11:27:41 PM Eastern European Standard Time
* Operating System: Version 11.1.2 (Build 15B202)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Friday, February 8, 2019 at 1:25:48 PM Eastern European Standard Time
* Operating System: Version 12.1 (Build 16B92)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

/*
* This header is generated by classdump-dyld 1.0
* on Sunday, November 10, 2019 at 11:16:35 PM Eastern European Standard Time
* Operating System: Version 13.1.3 (Build 17A878)
* Image Source: /System/Library/PrivateFrameworks/ControlCenterUIKit.framework/ControlCenterUIKit
* classdump-dyld is licensed under GPLv3, Copyright © 2013-2016 by Elias Limneos.
*/

//#import <ControlCenterUIKit/ControlCenterUIKit-Structs.h>
#import <UIKit/UIControl.h> // 11 // ?? 12,13 ??
//#import <UIKitCore/UIControl.h> // 12,13
//#import <UIKit/UIGestureRecognizerDelegate.h>

@class UIView, UIImageView, CCUICAPackageView, UIImage, UIColor, CCUICAPackageDescription, NSString;
// , MTVisualStylingProvider // 13

@interface CCUIButtonModuleView : UIControl <UIGestureRecognizerDelegate> {
	UIView* _highlightedBackgroundView;
	UIImageView* _glyphImageView;
//	CCUICAPackageView* _glyphPackageView;
//	MTVisualStylingProvider* _visualStylingProvider; // 13
	UIImage* _glyphImage;
	UIColor* _glyphColor;
	UIImage* _selectedGlyphImage;
	UIColor* _selectedGlyphColor;
	double _glyphAlpha; // 13
	double _glyphScale; // 13
//	CCUICAPackageDescription* _glyphPackageDescription;
	NSString* _glyphState;
//	NSDirectionalEdgeInsets _contentEdgeInsets;
}
@property (nonatomic,retain) UIImage * glyphImage;                                            //@synthesize glyphImage=_glyphImage - In the implementation block
@property (nonatomic,retain) UIColor * glyphColor;                                            //@synthesize glyphColor=_glyphColor - In the implementation block
@property (nonatomic,retain) UIImage * selectedGlyphImage;                                    //@synthesize selectedGlyphImage=_selectedGlyphImage - In the implementation block
@property (nonatomic,retain) UIColor * selectedGlyphColor;                                    //@synthesize selectedGlyphColor=_selectedGlyphColor - In the implementation block
@property (assign,nonatomic) double glyphAlpha; // 13                                              //@synthesize glyphAlpha=_glyphAlpha - In the implementation block
@property (assign,nonatomic) double glyphScale; // 13                                              //@synthesize glyphScale=_glyphScale - In the implementation block
//@property (nonatomic,retain) CCUICAPackageDescription * glyphPackageDescription;              //@synthesize glyphPackageDescription=_glyphPackageDescription - In the implementation block
@property (nonatomic,copy) NSString * glyphState;                                             //@synthesize glyphState=_glyphState - In the implementation block
//@property (assign,nonatomic) NSDirectionalEdgeInsets contentEdgeInsets;                       //@synthesize contentEdgeInsets=_contentEdgeInsets - In the implementation block
//@property (readonly) unsigned long long hash; 
//@property (readonly) Class superclass; 
//@property (copy,readonly) NSString * description; 
//@property (copy,readonly) NSString * debugDescription; 
-(void)setEnabled:(BOOL)arg1 ;
-(id)initWithFrame:(CGRect)arg1 ;
-(BOOL)gestureRecognizer:(id)arg1 shouldRecognizeSimultaneouslyWithGestureRecognizer:(id)arg2 ;
-(void)layoutSubviews;
-(void)setHighlighted:(BOOL)arg1 ;
-(void)didMoveToWindow; // 13
-(void)setSelected:(BOOL)arg1 ;
//-(void)setContentEdgeInsets:(NSDirectionalEdgeInsets)arg1 ;
//-(NSDirectionalEdgeInsets)contentEdgeInsets;
-(void)_handlePressGesture:(id)arg1 ;
-(UIColor *)selectedGlyphColor;
-(void)setSelectedGlyphColor:(UIColor *)arg1 ;
-(void)setGlyphColor:(UIColor *)arg1 ;
-(UIColor *)glyphColor;
-(UIImage *)glyphImage;
-(void)setGlyphImage:(UIImage *)arg1 ;
-(void)setSelectedGlyphImage:(UIImage *)arg1 ;
-(UIImage *)selectedGlyphImage;
-(void)_updateForStateChange;
//-(void)setGlyphPackageDescription:(CCUICAPackageDescription *)arg1 ;
-(void)setGlyphState:(NSString *)arg1 ;
//-(CCUICAPackageDescription *)glyphPackageDescription;
-(NSString *)glyphState;
-(void)setGlyphScale:(double)arg1 ; // 13
-(double)glyphScale; // 13
-(void)_setGlyphImage:(id)arg1 ;
-(void)_setGlyphPackageDescription:(id)arg1 ;
-(void)_setGlyphState:(id)arg1 ;
-(void)_setGlyphAlpha:(double)arg1 ; // 13
-(void)_setGlyphScale:(double)arg1 ; // 13
-(void)_updateGlyphImageViewVisualStyling; // 13
-(double)glyphAlpha; // 13
-(id)_tintColorForSelectedState:(BOOL)arg1 ; //13
-(void)setGlyphAlpha:(double)arg1 ; // 13
@end

